lib/puppet/util/windows/security.rb

```diff
@@ -372,7 +372,10 @@ def set_mode(mode, path, protected = true)
     dacl.allow(well_known_nobody_sid, nobody_allow)
 
     # TODO: system should be first?
-    dacl.allow(well_known_system_sid, system_allow)
+    flags = !isdir ? 0 :
+      Puppet::Util::Windows::AccessControlEntry::CONTAINER_INHERIT_ACE |
+      Puppet::Util::Windows::AccessControlEntry::OBJECT_INHERIT_ACE
+    dacl.allow(well_known_system_sid, system_allow, flags)
```

Comment:
If we don't allow inheritance, and the child (file/dir) is not something puppet is managing permissions on, `Dir.mkdir('/path/to/client_data/catalog')`, then this ACE won't apply to the child, and System can't write the cached catalog. Similarly, we don't manage the mode on the agent cert itself, e.g. `$ssldir\certs\foo.pem`, and System can't read the cert.

Also System is not normally the owner or group (that's only the case when you're running as System, e.g. as a service), then the CreatorOwner and CreatorGroup ACEs don't apply.